Transition 2.0 to 3.0

circulon opened the pull request:
Got the right Destination branch this time..

This PR includes all the features and fixes from the 2.0 branch
Also includes features from the current 3.x branch.

Additionally includes pre-commit setup as discussed in #935 
Linted and formatted the entire codebase as necessary.

@josephmancuso

josephmancuso committed:
black formatted

Commit:
Added method "only" to model class.

Commit:
updated only to support aliasing attribute names. This can be useful when mapping return data into new datasources which expect different naming conventions. This could be done at the database query level but is much easier to do using this method and only adds few extra LOC.

JarriqTheTechie committed:
Update Model.py

josephmancuso committed:
Update setup.py

josephmancuso committed:
fixed issue. needs refactoring

josephmancuso committed:
removed prints

josephmancuso committed:
fixed php ini

josephmancuso committed:
formatted

josephmancuso committed:
format

josephmancuso committed:
Merge branch 'features/only' of https://github.com/JarriqTheTechie/orm into JarriqTheTechie-features/only

josephmancuso committed:
added check for strings

josephmancuso committed:
added tests

josephmancuso committed:
Merge pull request #866 from JarriqTheTechie/features/only

Added only to Model

josephmancuso committed:
Merge pull request #868 from MasoniteFramework/fix/816

Fixed date not being able to be set to null

pmn4 committed:
BaseRelationship: s/self/instance

felipehertzer committed:
Mysql close the connection on rollback

amirhoseinsalimi committed:
fix: fix a typo in docs of `Collection.max`

amirhoseinsalimi committed:
feat: add `Collection.min` method

josephmancuso committed:
Merge pull request #870 from felipehertzer/patch-1

Close connection on rollback [MySQL]

josephmancuso committed:
Merge pull request #872 from amirhoseinsalimi/patch-1

fix: fix a typo in docs of `Collection.max`

josephmancuso committed:
Merge branch 'feature/collection-min' of https://github.com/amirhoseinsalimi/orm into amirhoseinsalimi-feature/collection-min

josephmancuso committed:
added min test

josephmancuso committed:
Merge pull request #874 from amirhoseinsalimi/feature/collection-min

feat: add `Collection.min` method

josephmancuso committed:
linted

josephmancuso committed:
Update setup.py

eaguad1337 committed:
wip: hasmany performance fix

josephmancuso committed:
fixed tests and made attempt

eaguad1337 committed:
fix: tests

eaguad1337 committed:
fix: foreign key call

josephmancuso committed:
removed print

josephmancuso committed:
Merge pull request #880 from MasoniteFramework/fix/relationship-performance-joe

Fix/relationship performance joe attempt

josephmancuso committed:
Update setup.py

josephmancuso committed:
Merge pull request #869 from pmn4/pmn4.base-relationship

BaseRelationship: s/self/instance

martinszemitis committed:
add configuration option to enable or disable foreign keys check for all connections

josephmancuso committed:
Merge pull request #883 from martinszemitis/2.0

Add configuration option to enable or disable foreign keys check for all 4 types of database connections

Kyrela committed:
fix: values in compile_alter_sql

Kyrela committed:
lint: remove additional blank line at end of file

Kyrela committed:
tests: `test_can_add_column_enum` and `test_can_change_column_enum`

Kyrela committed:
tests: `test_can_add_enum`

Kyrela committed:
fix: values in compile_alter_sql

josephmancuso committed:
fixed query builder

josephmancuso committed:
Merge pull request #885 from Kyrela/2.0

fix: values in compile_alter_sql

josephmancuso committed:
fixed has many though relationship

josephmancuso committed:
fixed has many

josephmancuso committed:
fixed querybuilder

josephmancuso committed:
fixed eager load

josephmancuso committed:
Merge pull request #890 from MasoniteFramework/hotfix/has-many

fixed has many

josephmancuso committed:
bumped version

josephmancuso committed:
Merge branch '2.0' of https://github.com/MasoniteFramework/orm into hotfix/fix-has_many_through

josephmancuso committed:
Merge pull request #889 from MasoniteFramework/hotfix/fix-has_many_through

fixed has many though relationship

josephmancuso committed:
bumped version

josephmancuso committed:
fixed missing where on has many through relationship

josephmancuso committed:
Merge pull request #891 from MasoniteFramework/hotfix/has-many-through-where

fixed missing where on has many through relationship

josephmancuso committed:
bumped version

josephmancuso committed:
fixed version

circulon committed:
use scopes helper method

circulon committed:
make QueryBuilder find respect scopes

circulon committed:
Added modles testts using scope

circulon committed:
consolidate duplicated calls

circulon committed:
Fixed queryBuilder .all() test

circulon committed:
fixed query=True not returning builder

Also remove redundant else clause

circulon committed:
use raw sql for clarity

circulon committed:
updated Model find and create

circulon committed:
moved find scope tests in with other tests

circulon committed:
Fixed HasOneThrough not loading distant model

Fixed eager and on demand loading of distant table
use vars for tablenames to increase readability.

circulon committed:
Fixed HasOneThrough with additional queries

circulon committed:
Fixed tests

josephmancuso committed:
Update pendulum version requirement

josephmancuso committed:
Update pendulum version requirement

josephmancuso committed:
Update pendulum version requirement

josephmancuso committed:
Merge pull request #898 from MasoniteFramework/feature/876-1

Upgrade to Pendulum 3

josephmancuso committed:
Update setup.py

circulon committed:
removed unused importsa and duplicate test

circulon committed:
Fixed linting

circulon committed:
Fixed typos

circulon committed:
Fixed eager loading

circulon committed:
Added tests for HasOneThrough eager loading

circulon committed:
removed unused import

circulon committed:
removed unused imports

circulon committed:
Fixed using .first()

fixed model attribute containing a list instead of a model when using .first()
added additional tests checking .first() usage

josephmancuso committed:
Refactor MySQLConnection to use connection pooling

josephmancuso committed:
Update version and dependencies in setup.py

josephmancuso committed:
Refactor PostgresConnection to use connection pooling

circulon committed:
fixed .on_null and .on_not_null had to be last criteria

circulon committed:
updated tests fo .on_null and .on_not_null

josephmancuso committed:
Merge pull request #901 from circulon/fix/on_null_always_last_criteria

fixed .on_null and .on_not_null had to be last criteria

josephmancuso committed:
Merge pull request #893 from circulon/fix/find_cannot_use_scopes

Fix find cannot use scopes

josephmancuso committed:
Merge pull request #895 from circulon/fix/has_one_through_not_working

Fix has one through relationship not working

josephmancuso committed:
Refactor MySQLConnection to use connection pooling

josephmancuso committed:
Refactor MySQLConnection to use connection pooling and add connection_pool_size parameter

josephmancuso committed:
Refactor MySQLConnection to use connection pooling and initialize connection pool with specified size

josephmancuso committed:
Refactor MySQLConnection to use connection pooling and remove debug print statements

josephmancuso committed:
Refactor MySQLConnection to use connection pooling and update connection pool size parameter

josephmancuso committed:
Refactor MySQLConnection to use connection pooling and update connection pool size parameter

josephmancuso committed:
Refactor MySQLConnection to use connection pooling and update connection pool size parameter

josephmancuso committed:
Refactor MySQLConnection to use connection pooling and update connection pool size parameter

josephmancuso committed:
format

josephmancuso committed:
Refactor PostgresConnection to use connection pooling and update connection pool size parameter

josephmancuso committed:
Refactor PostgresConnection to remove unused variable

josephmancuso committed:
Refactor PostgresConnection to enable connection pooling with a maximum size of 10

josephmancuso committed:
Refactor database configuration to update connection pool minimum size

josephmancuso committed:
Refactor SQLiteConnection to use a separate method for creating the connection

josephmancuso committed:
Refactor SQLiteConnection to use a separate method for creating the connection

josephmancuso committed:
linted

josephmancuso committed:
Refactor PostgresConnection to handle closed connections during query execution

josephmancuso committed:
Refactor PostgresConnection to handle closed connections during query execution

josephmancuso committed:
Refactor MySQLConnection to handle closed connections during query execution

josephmancuso committed:
Refactor MySQLConnection to handle closed connections during query execution

josephmancuso committed:
Refactor MySQLConnection to handle closed connections during query execution

josephmancuso committed:
linted

josephmancuso committed:
Refactor MySQLConnection to handle closed connections during query execution

josephmancuso committed:
Merge branch '2.0' into fix/query_true_should_return_query_builder

josephmancuso committed:
Merge pull request #892 from circulon/fix/query_true_should_return_query_builder

Tests query=True should return query builder

josephmancuso committed:
linted

circulon committed:
fixed HasManyThrough eager loading

circulon committed:
fixed HasManyThrough related model load

circulon committed:
cleaned up and fixed where clauses for HasManyThrough

circulon committed:
separated HasOneThrough and HasManyThrough tests

circulon committed:
fixed queriy tests

circulon committed:
make sure related attribute always contains a Collection

circulon committed:
Added ability so specify default select criteria for Models

Specify default select criteriaa for the Model by
overriding  get_selects() in the subclass

josephmancuso committed:
Merge pull request #904 from circulon/feature/902

Added ability so specify default select criteria for Models

josephmancuso committed:
Merge pull request #900 from MasoniteFramework/feature/483

Add connection pooling to MySQL and Postgres

josephmancuso committed:
Refactor PostgresConnection to support SSL options

josephmancuso committed:
Merge pull request #905 from MasoniteFramework/feature/875-1

Add PostgresConnection to support SSL and tls options

circulon committed:
use passed in parameters

circulon committed:
fixed parameter name

circulon committed:
renamed table variables

pbellerive committed:
fix nested relation

josephmancuso committed:
Add Radon analysis workflow for cyclomatic complexity checks

josephmancuso committed:
Enhance Radon analysis workflow to capture and output cyclomatic complexity results

josephmancuso committed:
Update Radon analysis workflow to save cyclomatic complexity output to a file and read from it

josephmancuso committed:
test

josephmancuso committed:
Update Radon analysis workflow to set output using GitHub environment variable

josephmancuso committed:
Update Radon analysis workflow to use set-output for environment variable

josephmancuso committed:
Refactor Radon analysis workflow to streamline output handling and remove redundant step

josephmancuso committed:
Update Radon analysis workflow to set output using GitHub environment variable

josephmancuso committed:
Update Radon analysis workflow to enhance reporting and streamline comments on pull requests

josephmancuso committed:
Update Radon comment action to use version 0.1

josephmancuso committed:
Fix version tag for Radon comment action in workflow

josephmancuso committed:
Add installation step for Radon in analysis workflow

josephmancuso committed:
Update Radon analysis workflow to report on changed files only

josephmancuso committed:
Update Radon analysis workflow to use the latest commit for reporting

josephmancuso committed:
Update Radon analysis workflow to use the correct base branch for reporting

josephmancuso committed:
Add test comment in __setitem__ method of Collection class

josephmancuso committed:
Fix Radon analysis workflow to use the correct branch reference for reporting

josephmancuso committed:
Update Radon analysis workflow to use the correct range for changed files

josephmancuso committed:
Handle empty file lists in Radon analysis workflow to prevent errors

josephmancuso committed:
Refactor Radon analysis workflow to support both push and pull_request events and ensure full git history is available

josephmancuso committed:
Simplify git fetch command in Radon analysis workflow to ensure full history is available

josephmancuso committed:
Update Radon analysis workflow to use the correct base branch for changed files

josephmancuso committed:
Refactor Radon analysis workflow to trigger on pull requests and streamline report generation

josephmancuso committed:
Update Radon analysis workflow to trigger on pull requests for the 2.0 branch

josephmancuso committed:
Add Radon installation step to analysis workflow

josephmancuso committed:
Add printing of Radon reports in analysis workflow

josephmancuso committed:
Enhance Radon analysis workflow to trigger on Python file changes and streamline report generation

josephmancuso committed:
Enhance Radon analysis workflow to fetch full git history and debug changed Python files

josephmancuso committed:
Refine Radon analysis workflow to improve environment variable handling and ensure proper output formatting

josephmancuso committed:
Add granular Radon analysis workflow for method-level change detection

josephmancuso committed:
Refine Radon analysis workflow to accurately match methods with changed lines and avoid duplicates

josephmancuso committed:
Refine Radon analysis workflow to capture raw output and improve method detection

josephmancuso committed:
Remove granular Radon analysis workflow configuration

josephmancuso committed:
wip

josephmancuso committed:
test

josephmancuso committed:
Remove commented test line and clean up migration command help text formatting

josephmancuso committed:
Merge pull request #908 from MasoniteFramework/feature/radon

Add Radon analysis workflow for cyclomatic complexity checks

josephmancuso committed:
Merge branch 'fix/has_many_through_not_working' of https://github.com/circulon/orm into circulon-fix/has_many_through_not_working

josephmancuso committed:
Merge pull request #903 from circulon/fix/has_many_through_not_working

Fix has many through relationship not working

josephmancuso committed:
Merge branch 'feature/906' of https://github.com/pbellerive/masonite_orm into pbellerive-feature/906

josephmancuso committed:
Refactor relationship mapping in QueryBuilder and related classes

josephmancuso committed:
Remove Radon analysis workflow from GitHub Actions

josephmancuso committed:
Add map_related method to HasMany relationship for grouping related results

josephmancuso committed:
Add map_related method to BelongsTo relationship for grouping related results

josephmancuso committed:
Add map_related method to HasOne and HasOneThrough relationships for grouping related results

josephmancuso committed:
Add map_related method to HasManyThrough relationship for grouping related results

josephmancuso committed:
linted

josephmancuso committed:
Fix missing newline at end of file in HasManyThrough relationship

josephmancuso committed:
Refactor map_related method in HasOne relationship to return related results directly

pbellerive committed:
remove group by

josephmancuso committed:
Merge branch 'feature/906' of https://github.com/pbellerive/masonite_orm into pbellerive-feature/906

josephmancuso committed:
Merge pull request #907 from pbellerive/feature/906

fix nested relation

josephmancuso committed:
Update setup.py

circulon committed:
Fixed using model default select columns correctly

Only use model get_selects() if no other columns are specified

circulon committed:
Fixed and added tests

circulon committed:
Cleanup

Removed unused imports
Removed commented out unused paramers
Sorted imports

josephmancuso committed:
Merge pull request #912 from circulon/fix/allow_override_of_model_default_selects

allow override of model default selects

josephmancuso committed:
Update setup.py

circulon committed:
Added where_not_in to Collection

circulon committed:
Check if args are all string ints first

No point in checking  the entire Collection to find no matches then cast the args to ints and test it again.

circulon committed:
fixed attach not working for unsaved models

Model attach_related() is now a pass-through for attach() (Model and BaseRelationship)
save_many() now uses attach()
Moved and fixed attach() logic to Baseelationship
This fixes an unsaved model could not be attached (especially importand for tables with mom-nullable foreign keys)

josephmancuso committed:
Merge pull request #918 from circulon/feature/collection_add_where_not_in

Add where_not_in to Collection

circulon committed:
moved eexception for not implemented methods to BaseRelationship

circulon committed:
added methods previously in BaseRelationship back into required classes

This is required as Relationships require different implementations.
Doing it this way helps encapsulate the Relationships

circulon committed:
sorted and black relationships

circulon committed:
Fixed typo

circulon committed:
aded attach/detach tests for HasOne and BelongsTo

circulon committed:
removed accidently duplicated method

circulon committed:
removed redundant parameters

circulon committed:
Added some comments

circulon committed:
Moved customisations to relationship classes

circulon committed:
changed nested pass logic in favour of fail early and readability

circulon committed:
enabled testing of increment, decrement and truncate queries

circulon committed:
fixed raw QueryBuilder could not use .find methods

Added an optional ‘colum’ parameter which is used when QueryBuilder has no model.
Fixed using a list of record ids generated an invalid wuery.

circulon committed:
added tests

circulon committed:
Fixed OFFSET with sqlite throwing syntax error

josephmancuso committed:
Merge pull request #930 from circulon/fix/929

Fixed OFFSET with sqlite throwing syntax error

josephmancuso committed:
Merge pull request #926 from circulon/fix/925

Fix 925

josephmancuso committed:
Merge pull request #928 from circulon/fix/927

Fix 927

josephmancuso committed:
Merge pull request #924 from circulon/fix/923

Fix 923

circulon committed:
Merge branch '2.0' of https://github.com/MasoniteFramework/orm into testing/2.0_to_3.0

# Conflicts:
#	.github/workflows/pythonapp.yml
#	.github/workflows/pythonpublish.yml
#	makefile
#	requirements.txt
#	setup.py
#	src/masoniteorm/connections/BaseConnection.py
#	src/masoniteorm/connections/MySQLConnection.py
#	src/masoniteorm/models/Model.py
#	src/masoniteorm/models/Model.pyi
#	src/masoniteorm/observers/ObservesEvents.py
#	src/masoniteorm/query/QueryBuilder.py
#	src/masoniteorm/query/grammars/MSSQLGrammar.py
#	src/masoniteorm/query/grammars/MySQLGrammar.py
#	src/masoniteorm/query/grammars/PostgresGrammar.py
#	src/masoniteorm/query/grammars/SQLiteGrammar.py
#	src/masoniteorm/schema/platforms/MySQLPlatform.py
#	src/masoniteorm/schema/platforms/PostgresPlatform.py
#	tests/mysql/builder/test_query_builder.py

circulon committed:
fixed having location in query string

circulon committed:
Fixed increment and decrement not returning builder

circulon committed:
setup pre-commit

make sure pre-commit is installed even if bootstrapping has been done previously

circulon committed:
Cleanup

circulon committed:
fixed attach on update

circulon committed:
fixed pre-commit install

circulon committed:
fixed CI workflows

circulon committed:
use pyproject settings

circulon committed:
fixed flak8 install

circulon committed:
removed requirements.dev dependency

circulon committed:
removed broken test

circulon committed:
Create and update default cast=True

includes black and isort formatting

circulon committed:
format

circulon committed:
Format & cleanup

circulon committed:
added tests to linting checks

circulon committed:
added tests to linting checks

circulon committed:
Merge branch 'feature/2.0_to_3.0' of https://github.com/circulon/orm into feature/2.0_to_3.0

circulon committed:
updated cleo version requiremnts

circulon committed:
updated compatibility versions

circulon committed:
Moved Faker to a dev/testing dependency

Faker is used for development and testing so will be installed via
‘make init’ and is not actually required in a production context

circulon committed:
moved psycopg2 driver check

circulon committed:
added seeder group to include the faker package

As Faker is only used for seeding this is appropriate use

circulon committed:
Added CHANGELOG

circulon commented:
@josephmancuso 
I think this is ready now ;)

Been testing it without issues ;)

Will have to update the docs with steps for upgrade
Luckily there are only a couple of things to address

circulon committed:
cleaned up casting

Fixed update not respecting casts
Removed redundant code duplication

circulon committed:
fixed bulk_create cast defaults to True

circulon committed:
only cast if a model is in use

circulon committed:
removed non passthrough method name

circulon committed:
aligned Model.pyi with Model and QueryBuilder

circulon committed:
removed duplicate method definition

circulon commented:
@josephmancuso  LGTM please

circulon committed:
reordered pre-commit tools

circulon committed:
Fixed Model autocompletes

Tagged all QueryBuilder “passthrough” methods as @classmethod
Added Hinting to make chaining easier
sorted methods by name

josephmancuso commented on src/masoniteorm/__init__.py:
```diff
@@ -1,2 +1 @@
 from .models import Model
-from .factories.Factory import Factory
```
why is this removed?

circulon replied:
The `Factory` class is only rewuired for Seeding as far as I can tell.

From a production environment perspective this is redundant
I added a [seeder] to the setup which will provide the Seeding capability if required.

I also noted that the Factory class should be imported from its package in the CHANGELOG in the  breaking changes section

josephmancuso commented on src/masoniteorm/commands/MigrateRefreshCommand.py:
```diff
@@ -33,11 +31,11 @@ def handle(self):
         if self.option("seed") == "null":
             self.call(
                 "seed:run",
-                f"None --directory {self.option('seed-directory')} --connection {self.option('connection', 'default')}",
```
Is this correct? Shouldn't this default to the `default` connection?

josephmancuso replied:
right now this would default to the None connection right?

circulon replied:
@josephmancuso 
unsure about this?
you altered it in commit 18de4662

circulon replied:
@josephmancuso 
seems to work fine from the command line for me 

Do you need to do some testing?
Or can we resolve this?

josephmancuso commented on src/masoniteorm/config.py:
```diff
@@ -12,8 +12,11 @@ def load_config(config_path=None):
     """
     selected_config_path = (
-        config_path or os.getenv("DB_CONFIG_PATH", None) or "config/database"
+        os.getenv("DB_CONFIG_PATH", None) or config_path or "config/database"
```
Change this to 

```python
os.getenv("DB_CONFIG_PATH", config_path) or "config/database"
```

josephmancuso replied:
its simpler like that

circulon committed:
Made selected config path decision easier to read

circulon commented:
@josephmancuso  LGTM

circulon committed:
Fixed some passthrough methods not hinted correctly

josephmancuso committed:
db

circulon committed:
Added comments for clarity of purpose

circulon committed:
Fixed Decimal not casting correctly

Decimal was not being output as a string when used for insert/update
Fixed tests

circulon committed:
updated orm.sqlite3 for tests

circulon commented:
@josephmancuso 

Can we move this forward please?

circulon committed:
added missing arguments

circulon commented:
@josephmancuso 
Is there anthing else that needs addressing to GTM?

josephmancuso committed:
Merge branch 'feature/2.0_to_3.0' of https://github.com/circulon/orm into circulon-feature/2.0_to_3.0
